ignite/metrics/gan/utils.py

```diff
@@ -0,0 +1,82 @@
+from typing import Any, Callable, Optional, Tuple, Union
+
+import torch
+
+from ignite.metrics.metric import Metric
+
+
+class InceptionModel(torch.nn.Module):
+    r"""Inception Model pre-trained on the ImageNet Dataset.
+
+    Args:
+        return_features: set it to `True` if you want the model to return features from the last pooling
+            layer instead of prediction probabilities.
+        device: specifies which device updates are accumulated on. Setting the
+            metric's device to be the same as your ``update`` arguments ensures the ``update`` method is
+            non-blocking. By default, CPU.
+    """
+
+    def __init__(self, return_features: bool, device: Union[str, torch.device] = "cpu") -> None:
+        try:
+            from torchvision import models
+        except ImportError:
+            raise RuntimeError("This module requires torchvision to be installed.")
+        super(InceptionModel, self).__init__()
+        self._device = device
+        self.model = models.inception_v3(pretrained=True).to(self._device)
+        if return_features:
+            self.model.fc = torch.nn.Identity()
+        self.model.eval()
+
+    @torch.no_grad()
+    def forward(self, data: torch.Tensor) -> torch.Tensor:
+        if data.dim() != 4:
+            raise ValueError(f"Inputs should be a tensor of dim 4, got {data.dim()}")
+        if data.shape[1] != 3:
+            raise ValueError(f"Inputs should be a tensor with 3 channels, got {data.shape}")
+        if data.device != torch.device(self._device):
+            data = data.to(self._device)
+        return self.model(data)
+
+
+class _BaseInceptionMetric(Metric):
+    def __init__(
+        self,
+        num_features: Optional[int] = None,
+        feature_extractor: Optional[torch.nn.Module] = None,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+    ) -> None:
+        if num_features is not None and num_features <= 0:
+            raise ValueError(f"Argument num_features must be greater to zero, got: {num_features}")
+
+        self._default_channels: int = 1
+        self._default_eval_model: Any = torch.nn.Identity
+        super(_BaseInceptionMetric, self).__init__(output_transform=output_transform, device=device)
+
+    def _check_input(
```

Comment:
Here, self._default_arg must be defined outside, before the call. It’s a bit unsafe.
Please, improve the design. An option would be to have this method being static.

Reply:
wouldn't the method have too many arguments in that case?

Reply:
I do appreciate the base metric for FID and IS but I'm not sure about the implementation. For instance, the _check_input is a factory rather a check.